### PR TITLE
Add gocql wrapper

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Add a gocql wrapper which will refresh session if ErrNoConnections occurs

**Why?**
Improve error handling & decrease operator intervention.

**How did you test it?**
N/A

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+//go:generate mockgen -copyright_file ../../../../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination interfaces_mock.go
+
+package gocql
+
+import (
+	"context"
+)
+
+// Note: this file defines the minimal interface that is needed by Cadence's cassandra
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/interfaces.go:33]: [Nit] Cadence vs Temporal in this comment?

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/interfaces.go:3]: Still okay to use year 2020 here?

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/query.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+	"github.com/gocql/gocql"
+)
+
+var _ Query = (*query)(nil)
+
+type (
+	query struct {
+		session *session
+		query   *gocql.Query
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/query.go:38]: same idea here....can we have the field name be differentiated from the struct name?

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+		batch:   b,
+	}
+}
+
+func (s *session) ExecuteBatch(
+	b Batch,
+) (retError error) {
+	defer func() { s.handleError(retError) }()
+
+	return s.Value.Load().(*gocql.Session).ExecuteBatch(b.(*batch).batch)
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go:138]: technically s.Value.Load() could return a different value than what was set on the original batch's session. Shouldn't be a problem, right?

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+	applied, iter, err := s.Value.Load().(*gocql.Session).MapExecuteBatchCAS(b.(*batch).batch, previous)
+	if iter == nil {
+		return applied, nil, err
+	}
+	return applied, iter, err
+}
+
+func (s *session) Close() {
+	session, ok := s.Value.Load().(*gocql.Session)
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go:155]: can this race with refresh() and cause refresh to create a new session after we closed the existing one?

**Review by mastermanu — APPROVED** [review]

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+	session := &session{
+		config:          config,
+		sessionInitTime: time.Now().UTC(),
+	}
+	session.Value.Store(gocqlSession)
+	return session, nil
+}
+
+func (s *session) refresh() error {
```
Comment (mastermanu) [inline common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go:73]: should we close the previous gocql session as well?

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go (hunk 4 above)
Comment (wxing1292) [reply common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go:138]: this one does not matter, as long as the thing coming out of atomic value is a `*gocql.Session`

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/batch.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+func (b *batch) WithContext(ctx context.Context) Batch {
+	b2 := b.batch.WithContext(ctx)
+	if b2 == nil {
+		return nil
+	}
+	return newBatch(b.session, b2)
+}
+
+func (b *batch) WithTimestamp(timestamp int64) Batch {
+	b.batch.WithTimestamp(timestamp)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/nosql/nosqlplugin/cassandra/gocql/batch.go:74]: some functions are side effect only, some are not..

File: common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go (hunk 5 above)
Comment (wxing1292) [reply common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go:155]: yes, there can be race causing session not to be closed.

**Pull request merged**